src/components/tl/mlx5/mcast/tl_mlx5_mcast_coll.c

```diff
@@ -155,16 +157,270 @@ static inline ucc_status_t ucc_tl_mlx5_mcast_do_bcast(ucc_tl_mlx5_mcast_coll_req
+                req->length, comm->max_per_packet);
+        return UCC_ERR_NOT_SUPPORTED;
+    }
 
-ucc_status_t ucc_tl_mlx5_mcast_test(ucc_tl_mlx5_mcast_coll_req_t* req)
+    if (req->mcast_prepost_bucket_size * req->concurrency_level * 2 > comm->params.rx_depth) {
+        tl_warn(comm->lib, "Pipelined mcast bcast not supported: "
+                "we only support the case prepost_bucket_size * concurrency_level * 2 > rx_depth, "
+                "but got: prepost_bucket_size=%d, concurrency_level=%d, "
+                "rx_depth=%d",
+                 req->mcast_prepost_bucket_size, req->concurrency_level,
+                 comm->params.rx_depth);
+        return UCC_ERR_NOT_SUPPORTED;
+    }
+
+    return UCC_OK;
+}
+
+static inline ucc_status_t
+ucc_tl_mlx5_mcast_prepare_zero_copy_bcast(ucc_tl_mlx5_mcast_coll_comm_t *comm,
+                                          ucc_tl_mlx5_mcast_coll_req_t  *req)
 {
-    ucc_status_t status = UCC_OK;
+    ucc_tl_mlx5_mcast_reg_t                   *reg    = NULL;
+    ucc_rank_t                                 root   = req->root;
+    int                                        offset = 0;
+    ucc_status_t                               status;
+    ucc_rank_t                                 j, i;
+    int                                        total_steps;
+    ucc_tl_mlx5_mcast_pipelined_ag_schedule_t *schedule;
+    ucc_assert(comm->bcast_comm.truly_zero_copy_bcast_enabled);
+
+    req->concurrency_level = comm->mcast_group_count / 2;
+    req->concurrency_level = ucc_min(req->concurrency_level, ONE_SIDED_MAX_CONCURRENT_LEVEL);
+    req->concurrency_level = ucc_min(req->concurrency_level, comm->commsize);
+    if (req->concurrency_level == 0) {
+        tl_warn(comm->lib, "not enough concurreny level to enable zcopy pipeline bcast");
+        return UCC_ERR_NOT_SUPPORTED;
+    }
 
-    ucc_assert(req->comm->psn >= req->start_psn);
+    req->mcast_prepost_bucket_size =
+        ucc_min(req->num_packets, comm->bcast_comm.mcast_prepost_bucket_size);
+
+    status = ucc_tl_mlx5_mcast_validate_zero_copy_bcast_params(comm, req);
+    if (status != UCC_OK) {
+        return status;
+    }
 
-    status = ucc_tl_mlx5_mcast_do_bcast(req);
+    /* calculate the schedule and details of what we should
+     * mcast and prepost to which mcast group at each stage*/
+    total_steps = req->num_packets / (req->concurrency_level
+                  * req->mcast_prepost_bucket_size) + 1;
+    schedule    = ucc_calloc(1, sizeof(ucc_tl_mlx5_mcast_pipelined_ag_schedule_t) *
+                             total_steps, "sched");
+    if (!schedule) {
+        tl_warn(comm->lib, "cannot allocate memory for schedule list");
+        return UCC_ERR_NO_MEMORY;
+    }
 
-    return status;
+    /* generate schedule */
+    for (i = 0; i < total_steps; i++) {
+        if (i < total_steps - 1) {
+            for (j = 0; j < req->concurrency_level; j++) {
+                schedule[i].prepost_buf_op[j].group_id =
```

Comment:
to be more readable - is it possible to move `schedule[i]` to separate variable `curr_schedule` (or similar)?
because it is easy to missprint somewhere i to j and hard to debug

Reply:
done